Instructions and Starter Code (#2)
lilykriopelle/codecademy-quiz-app: instructions.md

# Objective
This project – an app in which users can create and view topics and quizzes (which are comprised of flashcards) that are associated with those topics – gives you an opportunity to practice using Redux and Redux Toolkit to manage complex state.

## Functionality
On the `'/topics/new'` page:
  - Users can create topics
On the `'/topics'` page:
  – Users can view all topics
  – Users can click an individual topic and be redirected to the page for that topic
On the `/topics/:topicId` page:
  - Users can view an individual topic and all quizzes for that topic
  – Users can click a quiz associated with a topic and be redirected to that quiz's page
On the `'quizzes/new'` page:
  - Users can create quizzes that:
    - Are associated with topics
    – Contain lists of flashcards
      – Users can dynamically change the number of cards in a quiz by adding and removing card fields in the new quiz form
On the `'/quizzes'` page:
  - Users can view all quizzes
  – Users can click an individual quiz and be redirected to that quiz's page
On the `'/quizzes/:quizId'` page:
  - Users can view an individual quiz and flip cards over

>   – Users can delete a card associated with a quiz

Author: Gotcha. Nick asked for me to include descriptions of the interactions available on each page in the project objective in [his feedback here](https://codecademy.atlassian.net/browse/CDEV-626), but I can get rid of it if you think it's overkill.

>   – Users can delete a card associated with a quiz

Reviewer: I definitely think these descriptions should stay in the project but I think there's a better place than the narrative. Maybe this could go below the react-router dependency. Something like:

"At a high level, your application will be able to handle the following URL routes, each with their own functionality:

- On the `'/topics/new'` page:
    - Users can create topics
 - On the `'/topics'` page:
     - Users can view all topics
"

To get the list to render nicely, either make it a bulleted list with sub-bullets, or add line breaks before each "On the X page"

* Hint: Use `createSlice` to generate your topics slice. Your `addTopic` action should modify the `topics` object in state by adding an object representing a single topic to the `topics` object. Remember, we want the `topics` object to be keyed by topic `id`, so insert your newly created topic object as the value associated with the `id` you receive in the action's `payload`.

– Task: Add `topics` to the app's store.

* Hint: You'll need to import the reducer generated by your topics slice in **store.js** and add it to the empty object passed to `configureStore` under the key `topics`.

– Task: In **Topics.js**, import the selector defined in your slice and use it to access all the topics in state. Instead of mapping over an empty array, iterate over the topics in state, putting a `Link` on the page for each one. Likewise, in **Topic.js**, import the selector defined in your slice and use it to select the topics in state. We've written the code that pulls the topic's `id` out of the route params and uses it to select the correct topic from all the topics in state.

> * Hint: Inside the `<ul>` returned by the component, replace the empty array with topics stored in state, which you can access by using the selector you wrote in your topics slice. Because the topics are stored in an object keyed `id`, you will want to grab all of the object's values (using `Object.values()`) so that you can iterate over them using `map`.

Reviewer: It looks like this logic is already written out for them?

> – Task: Next, you'll need to hook the new topic form up to the action creators your slice generates. In **NewTopicForm.js**, import `addTopic` and dispatch it from the event handler that runs when the new topic form is submitted.  Verify that your code is working by filling out the form and submitting it. You should be redirected to the `/topics` page and should see your newly created topic there.

Reviewer: I think it would be good to have them import `useDispatch()` here too. What do you think?

* Hint: You will need to include the topic's `name` in the action payload as well as an `id` property (you should generate a value for this property by calling `uuidv4()`).

### Task Group 2: Quizzes
– Task: Great work! Now that you can create topics, your next task is to build out the necessary functionality to add quizzes to your app. This will involve creating two new slices—one for the quizzes themselves and one for the cards that comprise them—and adding an action to your topics slice to associate quizzes with the topic to which they belong. To start, create a slice for quizzes that:
  * Is named `'quizzes'`
  * Has initial state consisting of an object that includes one property, `quizzes`, which corresponds to an empty object. This object will eventually hold all quizzes keyed by `id`.
  * Has an `addQuiz` action. This action will receive a payload of the form `{ id: '123', name: 'quiz name', topicId: '456', cardIds: ['1', '2', '3', ...]}`.

> * Hint: Use `createSlice` to generate your quizzes slice. Your `addQuiz` action should modify the `quizzes` object in state by adding an object representing a single quiz to the `quizzes` object. Remember, we want the `quizzes` object to be keyed by quiz `id`, so insert your newly created quiz object as the value associated with the `id` you receive in the action's `payload`.

Reviewer: ```suggestion
Hint: 
* Use `createSlice` to generate your quizzes slice. 
* Your `addQuiz` action should modify the `quizzes` object in state by adding an object representing a single quiz to the `quizzes` object. 
* Remember, we want the `quizzes` object to be keyed by quiz `id`, so insert your newly created quiz object as the value associated with the `id` you receive in the action's `payload`.
```

– Task: Export the action creators and reducer your slice generates, as well as a selector that selects the `quizzes` object nested within `initialState` and add your quiz reducer to your app's store.

> * Hint: The action creators generated by your slice are accessible through the `actions` property on your slice, and the reducer is accessible via the `reducer` property.

Reviewer: When you implemented the Topics slice, this was included in the same task. I think you can do the same and combine this task with the previous task.

> – Task: Next, you should add an action to your topics slice that adds a quiz's `id` to the `quizIds` array of the topic with which the newly quiz is associated. This action action will receive a payload of the form `{quizId: '123', topicId: '456'}`.

Reviewer: They will also need to export the new action creator

* Hint: Use the payload's `topicId` to find the correct topic in state, and `push` the payload's `quizId` into that topic's `quizIds` array.

> – Task: Conceptually, these two actions are part of a single process: creating a new quiz and associating it with its topic. Now that you've written both actions, write an action creator that returns a thunk that dispatches them one after the other. This thunk action creator is the one that you will dispatch when a user creates a new quiz.

Reviewer: ```suggestion
– Task: Conceptually, creating a new quiz and associating it with its topic are part of a single process. Now that you've written both actions, write an action creator that returns a thunk that dispatches them one after the other. This thunk action creator is the one that you will dispatch when a user creates a new quiz.
```

Reviewer: I think you should specify that this is written in the quizzesSlice file

Reviewer: Also (this might go in the hints) I think you can remind them what the payload structure will be.

* Hint: Remember, thunks take the form:
```js
export const thunkActionCreator = (payload) => {
    return (dispatch) => {
      // dispatch multiple actions here
    };
};
```

> – Task: To test your work, you'll need to connect your action creator to **NewQuizForm** and make that component work. First, import your `topics` selector from your topics slice and replace our variable `topics` with a call to that selector.

Reviewer: ```suggestion
– Task: To test your work, you'll need to connect your action creator to **NewQuizForm** and make that component work. First, import your topics selector from your topics slice and replace assign a call to that selector to the variable `topics` (which is currently assigned an empty object).
```

* Hint: To use your selector you will need to call `useSelector` with the selector you defined in your topics slice.

> – Task: Next, import the thunk action creator from your quiz slice and dispatch it from the event handler that fires when the new quiz form is submitted. Remember, that action creator expects to receive a payload of the form `{ id: '123', name: 'quiz name', topicId: '456', cardIds: ['1', '2', '3', ...]}`. You'll have to generate an `id` by calling `uuidv4`. For now, pass an empty array for the `cardIds` property. Test that your action creator works by filling out the new quiz form. After your quiz is created you should be rerouted to the `/quizzes` page and should see your newly created quiz there.

Reviewer: ```suggestion
– Task: Next, import the thunk action creator from your quiz slice and dispatch it from the `handleSubmit()` event handler that fires when the new quiz form is submitted. 
* Remember, that action creator expects to receive a payload of the form `{ id: '123', name: 'quiz name', topicId: '456', cardIds: ['1', '2', '3', ...]}`. You'll have to generate an `id` by calling `uuidv4`. For now, pass an empty array for the `cardIds` property (you'll change that in a later task). 
* Test that your action creator works by filling out the new quiz form. After your quiz is created you should be rerouted to the `/quizzes` page and should see your newly created quiz there.
```

> * Hint: Your event handler should do something like:

Reviewer: ```suggestion
* Hint: Your dispatch statement should do something like:
```

```js
dispatch(
    addQuizForTopicId({
      name: '...',
      topicId: '...',
      cardIds: [],
      id: '...',
    })
);
```

> – Task: Lastly, import your selector in **Quizzes.js** and **Quiz.js** and make sure those components are displaying the correct data.

Reviewer: ```suggestion
– Task: Lastly, import your selector in **Quizzes.js** and **Quiz.js** and make sure those components are displaying the correct data: 
* The **Quizzes** component should render a **Link** for each quiz value in the quizzes slice of state.
* The **Quiz** component uses the `react-router-dom` method `useParams()` to determine the `quizId` to render. Therefore, it needs the full set of quizzes to find the appropriate quiz object to render.
```

– Hint: Use `useSelector` in conjunction with your selectors to pull the all the quizzes from state.

### Task Group 3: Cards
– Task: Great work! Next, create a slice for cards that:
* Is named `'cards'`
* Has initial state consisting of an object that includes one property, `cards`, which corresponds to an empty object. This object will eventually hold all cards keyed by `id`.
* Has an `addCard` action. This action will receive a payload of the form `{ id: '123', front: 'front of card', back: 'back of card'}`.

> * Has a `deleteCard` action that will receive a payload containing the `id` of the card to be deleted. Note: the best way to delete a property from an object is to use Javascript's built-in [`delete` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete).

Reviewer: The note about deleting a property from an object should go in the hints.

* Hint: Use `createSlice` to generate your cards slice. `addCard` should insert the payload into the `cards` object in state as the value associated with the `id` in the payload. `deleteCard` should remove the key and value corresponding to the `id` in the payload.

– Task: Export your slice's actions and reducer, and connect the cards reducer to your app's store.
* Hint: Add a `cards` key to the `reducers` object you pass to `configureStore` and set it equal to the reducer generated by your cards slice. Remember, that reducer is accessible via the slice's `reducer` property, and the action creators are accessible via the `actions` property.

> – Task: Lastly, connect your `addCard` action creator to the new quiz form. In **NewQuizForm**, in the event handler that fires when the quiz form is submitted, iterate through the `cards` in that form's local state, and for each one:
>   1. `dispatch` your `addCard` action creator. You will have to generate an `id` for each card using `uuidv4`.
>   2. Store the `id` you create for each card in an array.

Reviewer: What do you think about providing the cardIds array as an empty array?

```suggestion
– Task: Lastly, connect your `addCard` action creator to the new quiz form. In **NewQuizForm**, in the event handler that fires when the quiz form is submitted, iterate through the `cards` in that form's local state, and for each one:
  1. Generate an `id` for each card using `uuidv4`.
  2. Store that id in the `cardsId` array.
  3. `dispatch` your `addCard` action creator. 
```

> * Hint: Remember, your action creator expects to receive a payload of the form `{ id: '123', front: 'front of card', back: 'back of card'}`. You want to collect all the `cardIds` in an array so that you can pass them to the action creator that generates new quizzes.

Reviewer: Include in this hint and example of calling uuidv4()

> – Task: You previously passed an empty array for `cardIds` to the action creator that generates a new quiz. Now that you have  written code to collect an array of all the `cardIds` created whenever the new quiz form is submitted, replace the empty array with this array of `cardIds`. To test that your code is working, create a new quiz with some cards. Navigate to that quiz from the `/quizzes` page, and verify that your cards show up. Flip them over by clicking on them to make sure that you've correctly captured all of the state belonging to each card.

Reviewer: ```suggestion
– Task: You previously passed an empty array for `cardIds` to the action creator that generates a new quiz. Now that you have  written code to collect an array of all the `cardIds` created whenever the new quiz form is submitted, replace the empty array with this array of `cardIds`. 

To test that your code is working, create a new quiz with some cards. Navigate to that quiz from the `/quizzes` page, and verify that your cards show up. Flip them over by clicking on them to make sure that you've correctly captured all of the state belonging to each card.
```

* Hint: Since you're already dispatching the action that generates a new quiz with an appropriately structured payload, all you need to do is replace the empty array you previously assigned to  `cardIds` with an array that contains the `id`s belonging to the cards associated with the quiz that's being created.

> – Task: Now that you can add new cards, you'll need to display cards on the individual quiz page. The `Quiz` component renders a list of `Card` components, so in **Card.js**, import your cards selector and use it to access all the cards in state.

Reviewer: They also need to import the `useSelector` function

* Hint: Use `useSelector` in conjunction with your selector to get all the cards in state. We've already written the code that will use the `id` in props to select the relevant card.

– Task: Finally, import the action creator that deletes individual cards and connect it to the trash button that appears on each card.

> * Hint: Remember, that action creator expects to receive a payload of the form `{id: '123'}`. In the event handler that fires whenever a user clicks the trash button, dispatch your action creator with the appropriate payload.

Reviewer: Is this in NewQuizForm? If so, you should specify the file.

## Example Code
Want to see an example of how someone else has completed this project? Click this [link](TODO:REPLACE_ME) to download a zip file containing one example solution to this project. Remember: your project doesn't have to look anything like this! It should be unique to your vision.

## Sharing

Great work! Visit [our forums](TODO:REPLACE_ME) to compare your project to our sample solution code. After you host your own solution on GitHub, be sure to share it with other learners and see how other learners have built their own projects!

> Your solution might look different from ours, and that’s okay! There are multiple ways to solve these projects, and you’ll learn more by seeing others’ code.

Reviewer: These are great to have but are used in Portfolio projects, not Challenge projects. You can remove these sections and instead add in a Solutions task group as the third and final task group. Refer to the examples to see how this is done.

# Project Title

Quiz App

# Overview

> This project is slightly different than others you have encountered thus far on Codecademy. Instead of a step-by-step tutorial, this project contains a series of open-ended requirements which describe the project you’ll be building. There are many possible ways to correctly fulfill all of these requirements, and you should expect to use the internet, Codecademy, and other resources when you encounter a problem that you cannot easily solve.

Reviewer: Nick and I have been discussing removing this first sentence from all of these boilerplate overviews. We'd like to start by removing them from the projects in this skill path.

```suggestion
Instead of a step-by-step tutorial, this project contains a series of open-ended requirements which describe the project you’ll be building. There are many possible ways to correctly fulfill all of these requirements, and you should expect to use the internet, Codecademy, and other resources when you encounter a problem that you cannot easily solve.
```

  * Create a selector that selects the `topics` object nested within `initialState`.
  * Export the selector as well as the action creators and reducer that your slice generates.

Hint:
* Use `createSlice` to generate your topics slice. Your `addTopic` action should modify the `state.topics` object by adding an object representing a single topic to the `state.topics` object. Remember, we want the `state.topics` object to be keyed by the topic's `id` which you receive in the action's `payload`.
* You can use [computed property names](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#computed_property_names) to add values to an object using a variable as a key:

```js
const obj = {
  [key]: value,
}
```

– Task: Add `topics` to the app's store.

* Hint: You'll need to import the reducer generated by your topics slice in **store.js** and add it to the empty object passed to `configureStore` under the key `topics`.

– Task: In **Topics.js**, import the selector defined in your slice and use it to access all the topics in state, and replace the empty object currently assigned to `topics` with the topics in state.

> * Hint: Use `useSelector` conjunction with your selector to access all the topics in state.

Reviewer: ```suggestion
* Hint: You will need to import `useSelector` and then call it in conjunction with your selector to access all the topics in state.
```

– Task: Next, you'll need to hook the new topic form up to the action creators your slice generates. In **NewTopicForm.js**, import `addTopic` and dispatch it from the event handler that runs when the new topic form is submitted.  Verify that your code is working by filling out the form and submitting it. You should be redirected to the `/topics` page and should see your newly created topic there.

> * Hint: You will need to include the topic's `name` in the action payload as well as an `id` property (you should generate a value for this property by calling `uuidv4()`).

Reviewer: ```suggestion
* Hint: 
- You will need to import and call the `useDispatch()` method in order to dispatch actions to the store.
- You will need to include the topic's `name` in the action payload as well as an `id` property (you should generate a value for this property by calling `uuidv4()`).
```
